### PR TITLE
add ambient info to istiod config_dump API

### PR DESCRIPTION
**Pull request by GregHanson**
**Please provide a description of this PR:**

part of https://github.com/istio/istio/issues/42443

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ X ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

### CODE REVIEW

**Comment by GregHanson** [comment]
sample output:
```
k exec -it -n istio-system deploy/istiod -- curl -i localhost:15014/debug/config_dump?proxyID=ztunnel-k5q8p.istio-system
{
  "workloads": [
    {
      "Name": "sleep-66b946687b-5dbrj",
      "Namespace": "default",
      "Address": "10.244.1.5",
      "Identity": "spiffe://cluster.local/ns/default/sa/sleep",
      "Node": "ambient-worker2"
    },
    {
      "Name": "helloworld-v2-same-node-65bd845857-drk2r",
      "Namespace": "default",
      "Address": "10.244.1.4",
      "Identity": "spiffe://cluster.local/ns/default/sa/helloworld",
      "Node": "ambient-worker2"
    },
    {
      "Name": "helloworld-v1-cross-node-55446d46d8-b2snz",
      "Namespace": "default",
      "Address": "10.244.2.7",
      "Identity": "spiffe://cluster.local/ns/default/sa/helloworld",
      "Node": "ambient-worker"
    },
    {
      "Name": "swiss-army-knife-58c658bdfb-chmpl",
      "Namespace": "default",
      "Address": "10.244.2.9",
      "Identity": "spiffe://cluster.local/ns/default/sa/sleep",
      "Node": "ambient-worker"
    }
  ],
  "waypoints": [
    {
      "Name": "helloworld-waypoint-proxy-db756678b-dtw9h",
      "Namespace": "default",
      "Address": "10.244.2.6",
      "Identity": "spiffe://cluster.local/ns/default/sa/helloworld",
      "Node": "ambient-worker"
    }
  ],
  "ztunnels": [
    {
      "Name": "ztunnel-wfxvs",
      "Namespace": "istio-system",
      "Address": "10.244.1.3",
      "Identity": "spiffe://cluster.local/ns/istio-system/sa/ztunnel",
      "Node": "ambient-worker2"
    },
    {
      "Name": "ztunnel-k5q8p",
      "Namespace": "istio-system",
      "Address": "10.244.0.5",
      "Identity": "spiffe://cluster.local/ns/istio-system/sa/ztunnel",
      "Node": "ambient-control-plane"
    },
    {
      "Name": "ztunnel-xzcrp",
      "Namespace": "istio-system",
      "Address": "10.244.2.2",
      "Identity": "spiffe://cluster.local/ns/istio-system/sa/ztunnel",
      "Node": "ambient-worker"
    }
  ]
}
```

**Comment by GregHanson** [comment]
@howardjohn is there any difference between using the workloads stored in `conn.proxy.LastPushContext.AmbientIndex` versus [WorkloadGenerator.GenerateDeltas()](https://github.com/istio/istio/blob/experimental-ambient/pilot/pkg/xds/workload.go#L45)

File: pilot/pkg/xds/debug.go [hunk 1]
```diff
@@ -687,11 +688,20 @@ func (s *DiscoveryServer) ConfigDump(w http.ResponseWriter, req *http.Request) {
 		return
 	}
 	if ts := s.getResourceTypes(req); len(ts) != 0 {
+		if con.proxy.IsZTunnel() {
+			istiolog.Warnf("proxy %v is not envoy based", proxyID)
```
Comment (howardjohn) [inline pilot/pkg/xds/debug.go:692]: do we need this?

File: pilot/pkg/xds/debug.go [hunk 2]
```diff
@@ -687,11 +688,20 @@ func (s *DiscoveryServer) ConfigDump(w http.ResponseWriter, req *http.Request) {
...
+		}
 		dump := s.getConfigDumpByResourceType(con, ts)
 		writeJSON(w, dump, req)
 		return
 	}
 
+	if con.proxy.IsZTunnel() {
+		workloads := s.debugWorkloads(con.proxy.LastPushContext.AmbientIndex)
```
Comment (howardjohn) [inline pilot/pkg/xds/debug.go:700]: I think we want to return the result of the workload generator (using getConfigDumpByResourceType L694). This part is already exposed  under ambientz and is mostly useless for ztunnel (only used for waypoint)

File: pilot/pkg/xds/debug.go (hunk 1 above)
Comment (GregHanson) [reply pilot/pkg/xds/debug.go:692]: was curious about this case -  requesting things like listeners/cluster/routes/etc through `types` query params are not going to apply to ztunnels and could be confusing to users if they're expecting them to work. WDYT?

File: pilot/pkg/xds/debug.go (hunk 2 above)
Comment (GregHanson) [reply pilot/pkg/xds/debug.go:700]: any thoughts one way or the other about returning the entire workload vs the simplified struct like in ambientz?  The ztunnel config_dump API is already returning a different internal struct, so the outputs cannot be diffed directly to see what's missing

File: pilot/pkg/xds/debug.go (hunk 1 above)
Comment (howardjohn) [reply pilot/pkg/xds/debug.go:692]: This is a low level debugging interface, IMO we should just return what they ask. Users aren't going to see the istiod logs anyways, and its perfectly valid for ztunnel's to call this (and not put types as listners, etc)

File: pilot/pkg/xds/debug.go (hunk 2 above)
Comment (howardjohn) [reply pilot/pkg/xds/debug.go:700]: /config_dump should take in a list of types and a proxy and return the XDS for that type+proxy combination.

**Comment by GregHanson** [comment]
after discussion, created PR to master branch to allow arbitrary types as query params to the /config_dump API: https://github.com/istio/istio/pull/42658

Will merge master branch PR first, and update this PR with the changes

**Comment by linsun** [comment]
> 

What is the benefit to users for the list of ztunnels in this return? Also, how can user easily figure out if a producer has waypoint?

**Comment by howardjohn** [comment]
> Re your sample output earlier: [#42547 (comment)](https://github.com/istio/istio/pull/42547#issuecomment-1362185385)
> 
> What is the benefit to users for the list of ztunnels in this return? Also, how can user easily figure out if a producer has waypoint?

The purpose of the config_dump tool is to dump the XDS config as-is,  not to manipulate it in a way that may be useful to users. Its a low level debugging command

File: pilot/pkg/xds/debug.go (hunk 1 above)
Comment (linsun) [reply pilot/pkg/xds/debug.go:692]: yeah, not sure why users would want to see these ztunnel infos.

File: pilot/pkg/xds/debug.go (hunk 2 above)
Comment (linsun) [reply pilot/pkg/xds/debug.go:700]: yeah, ideally, formatted similarly to what the ztunnel receives so we can easily compare the configs (from istiod vs ztunnel).

**Comment by linsun** [comment]
> > Re your sample output earlier: [#42547 (comment)](https://github.com/istio/istio/pull/42547#issuecomment-1362185385)
> > What is the benefit to users for the list of ztunnels in this return? Also, how can user easily figure out if a producer has waypoint?
> 
> The purpose of the config_dump tool is to dump the XDS config as-is, not to manipulate it in a way that may be useful to users. Its a low level debugging command

Ok fair, since this is lower level debug. I guess I would expect similar output as I dump the config on the ztunnel ((using istioctl pc workload ztunnel cmd) so I can compare easily.

File: pilot/pkg/xds/debug.go [hunk 3]
```diff
@@ -842,18 +829,28 @@ func (s *DiscoveryServer) configDump(conn *Connection, includeEds bool) (*admin.
 		return nil, err
 	}
 
+	// extentionsConfig := make([]*admin.EcdsConfigDump_EcdsFilterConfig, 0)
```
Comment (GregHanson) [inline pilot/pkg/xds/debug.go:832]: this is dependent on an update to the go-control-plane library

**Comment by GregHanson** [comment]
sample config_dump output:
```
{
  "configs": [
    {
      "@type": "type.googleapis.com/istio.workload.Workload",
      "name": "istio-ingressgateway-5b54f56fb5-45wvp",
      "namespace": "istio-system",
      "address": "CvQBAw==",
      "serviceAccount": "istio-ingressgateway-service-account",
      "node": "ambient-worker",
      "canonicalName": "istio-ingressgateway",
      "canonicalRevision": "latest",
      "workloadName": "istio-ingressgateway",
      "virtualIps": {
        "10.96.16.71": {
          "ports": [
            {
              "servicePort": 15021,
              "targetPort": 15021
            },
            {
              "servicePort": 80,
              "targetPort": 8080
            },
            {
              "servicePort": 443,
              "targetPort": 8443
            }
          ]
        }
      }
    },
    {
      "@type": "type.googleapis.com/istio.workload.Workload",
      "name": "helloworld-v1-cross-node-55446d46d8-mhclj",
      "namespace": "default",
      "address": "CvQBBw==",
      "protocol": "HTTP",
      "serviceAccount": "helloworld",
      "waypointAddresses": [
        "CvQBBg=="
      ],
      "node": "ambient-worker",
      "canonicalName": "helloworld",
      "canonicalRevision": "v1",
      "workloadName": "helloworld-v1-cross-node",
      "virtualIps": {
        "10.96.52.152": {
          "ports": [
            {
              "servicePort": 5000,
              "targetPort": 5000
            }
          ]
        },
        "10.96.87.114": {
          "ports": [
            {
              "servicePort": 5000,
              "targetPort": 5000
            }
          ]
        }
      }
    },
    {
      "@type": "type.googleapis.com/istio.workload.Workload",
      "name": "sleep-66b946687b-4xg8x",
      "namespace": "default",
      "address": "CvQCBQ==",
      "protocol": "HTTP",
      "serviceAccount": "sleep",
      "node": "ambient-worker2",
      "canonicalName": "sleep",
      "canonicalRevision": "latest",
      "workloadName": "sleep",
      "virtualIps": {
        "10.96.216.244": {
          "ports": [
            {
              "servicePort": 80,
              "targetPort": 80
            }
          ]
        }
      }
    },
    {
      "@type": "type.googleapis.com/istio.workload.Workload",
      "name": "local-path-provisioner-9cd9bd544-c6dv6",
      "namespace": "local-path-storage",
      "address": "CvQABA==",
      "serviceAccount": "local-path-provisioner-service-account",
      "node": "ambient-control-plane",
      "canonicalName": "local-path-provisioner",
      "canonicalRevision": "latest",
      "workloadName": "local-path-provisioner"
    },
    {
      "@type": "type.googleapis.com/istio.workload.Workload",
      "name": "coredns-6d4b75cb6d-55wss",
      "namespace": "kube-system",
      "address": "CvQAAw==",
      "serviceAccount": "coredns",
      "node": "ambient-control-plane",
      "canonicalName": "coredns",
      "canonicalRevision": "latest",
      "workloadName": "coredns",
      "virtualIps": {
        "10.96.0.10": {
          "ports": [
            {
              "servicePort": 53,
              "targetPort": 53
            },
            {
              "servicePort": 9153,
              "targetPort": 9153
            }
          ]
        }
      }
    },
    {
      "@type": "type.googleapis.com/istio.workload.Workload",
      "name": "istiod-747687bbb6-kzm2k",
      "namespace": "istio-system",
      "address": "CvQCAg==",
      "serviceAccount": "istiod",
      "node": "ambient-worker2",
      "canonicalName": "istiod",
      "canonicalRevision": "latest",
      "workloadName": "istiod",
      "virtualIps": {
        "10.96.82.77": {
          "ports": [
            {
              "servicePort": 15010,
              "targetPort": 15010
            },
            {
              "servicePort": 15012,
              "targetPort": 15012
            },
            {
              "servicePort": 443,
              "targetPort": 15017
            },
            {
              "servicePort": 15014,
              "targetPort": 15014
            }
          ]
        }
      }
    },
    {
      "@type": "type.googleapis.com/istio.workload.Workload",
      "name": "ztunnel-4h2wj",
      "namespace": "istio-system",
      "address": "CvQCAw==",
      "serviceAccount": "ztunnel",
      "node": "ambient-worker2",
      "canonicalName": "ztunnel",
      "canonicalRevision": "latest",
      "workloadType": "POD",
      "workloadName": "ztunnel-4h2wj"
    },
    {
      "@type": "type.googleapis.com/istio.workload.Workload",
      "name": "ztunnel-zplsv",
      "namespace": "istio-system",
      "address": "CvQBAg==",
      "serviceAccount": "ztunnel",
      "node": "ambient-worker",
      "canonicalName": "ztunnel",
      "canonicalRevision": "latest",
      "workloadType": "POD",
      "workloadName": "ztunnel-zplsv"
    },
    {
      "@type": "type.googleapis.com/istio.workload.Workload",
      "name": "sleep2-7886c67bf8-xf6hv",
      "namespace": "default",
      "address": "CvQBCA==",
      "serviceAccount": "sleep",
      "node": "ambient-worker",
      "canonicalName": "sleep2",
      "canonicalRevision": "latest",
      "workloadName": "sleep2"
    },
    {
      "@type": "type.googleapis.com/istio.workload.Workload",
      "name": "swiss-army-knife-58c658bdfb-22q8n",
      "namespace": "default",
      "address": "CvQBCQ==",
      "protocol": "HTTP",
      "serviceAccount": "sleep",
      "node": "ambient-worker",
      "canonicalName": "swiss-army-knife",
      "canonicalRevision": "latest",
      "workloadName": "swiss-army-knife"
    },
    {
      "@type": "type.googleapis.com/istio.workload.Workload",
      "name": "coredns-6d4b75cb6d-pp5c8",
      "namespace": "kube-system",
      "address": "CvQAAg==",
      "serviceAccount": "coredns",
      "node": "ambient-control-plane",
      "canonicalName": "coredns",
      "canonicalRevision": "latest",
      "workloadName": "coredns",
      "virtualIps": {
        "10.96.0.10": {
          "ports": [
            {
              "servicePort": 53,
              "targetPort": 53
            },
            {
              "servicePort": 9153,
              "targetPort": 9153
            }
          ]
        }
      }
    },
    {
      "@type": "type.googleapis.com/istio.workload.Workload",
      "name": "ztunnel-5bjfw",
      "namespace": "istio-system",
      "address": "CvQABQ==",
      "serviceAccount": "ztunnel",
      "node": "ambient-control-plane",
      "canonicalName": "ztunnel",
      "canonicalRevision": "latest",
      "workloadType": "POD",
      "workloadName": "ztunnel-5bjfw"
    },
    {
      "@type": "type.googleapis.com/istio.workload.Workload",
      "name": "helloworld-v2-same-node-65bd845857-vld2n",
      "namespace": "default",
      "address": "CvQCBA==",
      "protocol": "HTTP",
      "serviceAccount": "helloworld",
      "waypointAddresses": [
        "CvQBBg=="
      ],
      "node": "ambient-worker2",
      "canonicalName": "helloworld",
      "canonicalRevision": "v2",
      "workloadName": "helloworld-v2-same-node",
      "virtualIps": {
        "10.96.133.109": {
          "ports": [
            {
              "servicePort": 5000,
              "targetPort": 5000
            }
          ]
        },
        "10.96.87.114": {
          "ports": [
            {
              "servicePort": 5000,
              "targetPort": 5000
            }
          ]
        }
      }
    }
  ]
}
```

**Comment by GregHanson** [comment]
verified that the modified `/config_dump` API does produce different results than `istioctl pc workload` prior to John's fix for this issue [here](https://github.com/istio/istio/pull/42682)

**Review by howardjohn — APPROVED** [review]

File: pilot/pkg/xds/debug.go (hunk 3 above)
Comment (GregHanson) [reply pilot/pkg/xds/debug.go:832]: update PR: https://github.com/istio/istio/pull/42702

**Pull request merged**